### ISSUE TITLE
[FEATURE] FLOOD CONTROL

### ISSUE THREAD

**Issue by Gitcqs** [python-telegram-bot/python-telegram-bot#2362]
Regarding flow control, my robot was recently attacked by a similar ddos ​​and was forced to make a large number of responses in a short period of time. The following problems were found using messagequeue:
1. The response of normal users will be completely blocked by the attacker's response message for a period of time before responding
2. Different telegram messages actually have different flow control strategies. According to my test, when a message larger than about 512 bytes prompts hit limit, short messages can still be sent.

My solution
1. Do not respond to some messages discarded and prompt the user. At the same time, the message needs a priority queue, because normal user message response will be blocked for too long.
2. In fact, it is not quite sure how to deal with it. If the 1 strategy is conservative, it should be solved, otherwise the response message can only be controlled by size detection.

It is expected that the flow control module can add the following features:
1: One step closer to distinguish large messages for traffic restriction
2: Add an option to provide a priority queue to prevent normal user messages from being blocked, and give priority to responding to new users in the queue, as for the remaining messages to be released slowly, so as to prevent the linear queue from causing a particularly poor user experience

**Comment by Bibo-Joshi** [comment]
Hi. I think we're talking about two different things here:

1. How to prevent users from spamming the bot
2. How to avoid hitting flood limits

IMO effective spam protection would prevent spammy input leading to requests to TG (which would then go to the `MessageQueue`) in the first place. That way the updates from non-spam users are handled normally by the `MessageQueue`.

This can be achieved by putting a `TypeHandler(telegram.Update, rate_limiting_callback)` in a low group. When spam is detected, respond to it appropriately and raise `DipsatcherHandlerStop`. See also [here](https://t.me/pythontelegrambotgroup/397190). Ofc the rate limiting logic can use something like a `DelayQueue`. Please keep in mind that `telegram.ext.DelayQueue` is intended for the internal use in `MessageQueue` only and will likely be refactored or even removed at some point in the context of #2139 .

However such a spam protection feature is highly customized and out of scope for PTB imo. Good thing we introduced ptbcontrib a while ago :) As said in the message linked above, I do think that such a thing would be a good fit for ptbcontrib!

Regarding fine tuning of `MessegaeQueues` limits: `MQ` is still in a very rough state from the initial addition a few years back. We have plans to rework it in a way that allows for more customization (see #2139 mentioned above), but that will have to wait for v14.

**Comment by Gitcqs** [comment]
Thanks for your reply, the suggested method will be adopted.
Hope that the problem can be solved directly😊.

(Otherwise, according to the buried log, the delay queue is too long and the response of normal users is too slow. The message delay queue needs to be changed from queue (FIFO) to PriorityQueue, and the priority is based on the user's unresponsive time. This is very similar to the process scheduling algorithm. FIFO leads to process starvation and requires other scheduling algorithms. I hope I don't need such a complicated and inelegant strategy.)

**Comment by Bibo-Joshi** [comment]
I'm not sure if you're talking about spam prevention or flood limits now.

If you're talking about spam prevention: As sad above, `DelayQueue` is intended for the use within `MessageQueue` only. If you want to somehow use a priority queue for spam prevention, please implement your own logic for that.

If you're talking about flood limits: As pointed out, refactoring of `MessageQueue` is on the post-v14-ToDo list and extending the functionality to usa a priorty logic was indeed already proposed, see #2265 :)

**Comment by Gitcqs** [comment]
<img width="250" alt="截屏2021-02-06 下午6 01 44" src="https://user-images.githubusercontent.com/12613289/107115135-68148580-68a5-11eb-8165-a58ed78af436.png">

Yes, it is the response priority (the problem associated with spam and flood limit). According to my test, a simple click script (if the robot does not have a spam mechanism), even after a short period of accumulation, It will make the robot block the normal response for 10 seconds-20 seconds (because the robot must avoid the flood limit. At the same time, if there is no spam mechanism, it will respond to every message, disaster will happen).
Roughly implemented a thread-safe queue with dynamic priority, finally solved the problem that the troublemaker continued to force the robot to make massive responses. The normal response blocking dropped from 20 seconds to 1 second. MessageQueue absolutely needs priority logic, because most of the current robots do not have a spam mechanism, and a click script can completely ruin the robot interaction experience.

Thanks my problem is solved❤️

 Replace queue.Queue
`
class Dynamic_priority_queue:
    
    def __init__(self):
        self.li = []
        self.lock = threading.Lock()
        self.__mutex = threading.Lock()
        self.__cond = threading.Condition(self.__mutex)
        self.last_response = {}

    def get(self):
        if self.__cond.acquire():
            while not self.li:
                self.__cond.wait()

            item = self.li.pop()
            self._mark(item, "get")
            self._sort()

            self.__cond.notify()
            self.__cond.release()
        return item

    def put(self, item):
        if self.__cond.acquire():
            self.li.append(item)
            self._mark(item, "put")
            self._sort()

            self.__cond.notify()
            self.__cond.release()

    def _sort(self):
        self.li.sort(key=lambda x: self.last_response[self._chat_id(x)], reverse=True)

    def _mark(self, item, action):
        chat_id = self._chat_id(item)
        if action == "put":
            if chat_id not in self.last_response:
                self.last_response[chat_id] = time.time()
                return True
        else:
            self.last_response[chat_id] = time.time()
            return True

    def _chat_id(self, item):
        chat_id = item[0].kwargs.get("chat_id")
        if chat_id is None:
            chat_id = item[0].args[1]
        return chat_id
`

**Comment by Bibo-Joshi** [comment]
I'm not sure if I can follow, but this again reads like you want to change how `MessageQueue` processes the requests. I'll repeat that

> IMO effective spam protection would prevent spammy input leading to requests to TG (which would then go to the MessageQueue) in the first place. That way the updates from non-spam users are handled normally by the MessageQueue.

But if your solution works fine for you, that's ofc okay and due to

> Thanks my problem is solved❤️

I'll go ahead and close this :)

**Comment by Gitcqs** [comment]
> > IMO effective spam protection would prevent spammy input leading to requests to TG (which would then go to the MessageQueue) in the first place. That way the updates from non-spam users are handled normally by the MessageQueue.

Okay, I will try to add spam detection.
